src/DynamoCore/Graph/Nodes/NodeModel.cs

```diff
@@ -668,6 +681,18 @@ private bool CheckIfAnyUpstreamNodeIsFrozen(NodeModel node, ref bool ret)
 
             return ret;
         }
+
+        private void MarkDownStreamNodesAsModified(NodeModel node)
+        {                         
```

Comment:
It's extremely important to avoid duplicate `NodeModified` events.

It is more efficient to maintain a hash set of "already marked" nodes. You could reuse your more general algorithm (`GetDownstreamNodes`), which is currently on `EngineController`. 

So, I'd recommend moving your `GetDownstreamNodes` method to `NodeModel` and just do.

```csharp
var nodes = GetDownstreamNodes();
foreach ( var node in nodes ) {
   // mark modified
}
```